frameworklauncher/src/main/java/com/microsoft/frameworklauncher/zookeeperstore/ZooKeeperClient.java

```diff
@@ -143,19 +147,50 @@ public void deleteRecursively(String path, Boolean childrenOnly) throws Exceptio
   // Set/Get small size (<= ZK_MAX_NODE_BYTES) yaml object to the node of the given path, no matter the given path exist or not.
   // DISTRIBUTED THREAD SAFE
   public <T> void setSmallYamlObject(String path, T yamlObject) throws Exception {
-    createNode(path, YamlUtils.toBytes(yamlObject));
+    byte[] serializedObject = YamlUtils.toBytes(yamlObject);
+
+    long start = System.currentTimeMillis();
+
+    byte[] payload;
+    if (zkCompressionEnable) {
+      payload = CompressionUtils.compress(serializedObject);
+    } else {
+      payload = serializedObject;
+    }
+    createNode(path, payload);
```

Comment:
>createNode(path, payload); [](start = 4, length = 26)

Have you ever considered add compress/decompress functionality into the lowerest zookeeper write/read functions with a flag parameter? By this way you can save code if other places also needs compression/decompression.

Reply:
Yes, that is another approach. However, to achieve better compression, we take the whole large object as the compression input instead of take each part of the whole large object as the compression input.
Pls check line 230.

---
In reply to: [170508297](https://github.com/Microsoft/pai/pull/338#discussion_r170508297) [](ancestors = 170508297)